### PR TITLE
Reintroduce separate lifetimes for buffer and stream

### PR DESCRIPTION
**Pull request by bugadani**

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -326,10 +326,10 @@ where
     /// The response headers are stored in the provided rx_buf, which should be sized to contain at least the response headers.
     ///
     /// The response is returned.
-    pub async fn send<'buf>(
-        &'buf mut self,
+    pub async fn send<'buf, 'req>(
```
Comment (rmja) [inline src/client.rs:329]: Can we swap the lifetime arguments here to align with the other send functions. We have throughout tried to have the convention that the longest lifetimes "comes first"

File: src/client.rs [hunk 2]
```diff
@@ -454,11 +454,11 @@ where
     /// The response headers are stored in the provided rx_buf, which should be sized to contain at least the response headers.
     ///
     /// The response is returned.
-    pub async fn send<'req, B: RequestBody>(
+    pub async fn send<'buf, 'req, B: RequestBody>(
```
Comment (rmja) [inline src/client.rs:457]: Same thing here

File: tests/client.rs [hunk 3]
```diff
@@ -395,3 +395,18 @@ async fn echo(req: hyper::Request<Body>) -> Result<hyper::Response<Body>, hyper:
...
+fn compile_tests() {
+    #[allow(dead_code)]
+    async fn post_ping<'buf>(port: u16, rx_buf: &'buf mut [u8]) -> &'buf mut [u8] {
+        let mut http = HttpClient::new(&TCP, &LOOPBACK_DNS);
+        let url = format!("http://127.0.0.1:{}", port);
+        let mut request = http.request(Method::GET, &url).await.unwrap();
+        let response = request.send(rx_buf).await.unwrap();
+        if !response.status.is_successful() {
+            panic!("Firmware download failed, status code {:?}", response.status);
+        }
```
Comment (rmja) [inline tests/client.rs:409]: Please remove these three lines, they are very specific to my app:)

File: tests/client.rs [hunk 4]
```diff
@@ -395,3 +395,18 @@ async fn echo(req: hyper::Request<Body>) -> Result<hyper::Response<Body>, hyper:
         _ => Ok(hyper::Response::new(req.into_body())),
     }
 }
+
+#[test]
+fn compile_tests() {
+    #[allow(dead_code)]
+    async fn post_ping<'buf>(port: u16, rx_buf: &'buf mut [u8]) -> &'buf mut [u8] {
```
Comment (rmja) [inline tests/client.rs:402]: Maybe rename to "buffer_lifetime" or something like it

**Review by rmja — APPROVED** [review]

File: tests/client.rs (hunk 4 above)
Comment (bugadani) [reply tests/client.rs:402]: Sorry for pasting your own repro in verbatim :P

**Comment by rmja** [comment]
Perfect!

**Pull request merged**